# Multiple Linear Regression ToyotaCorolla Problem Statement
Problem Statements : 
An online car sales platform would like to improve its customer base and their experience by providing them an easy way to buy and sell cars. For this, they would like an automated model which can predict the price of the car once the user inputs the required factors. Help the business achieve their objective by applying multilinear regression on the given dataset. Please use the below columns for the analysis purpose: price, age_08_04, KM, HP, cc, Doors, Gears, Quarterly_Tax, and Weight.

Business Objectives : To predict the price of a car based on there  price, age_08_04, KM, HP, cc, Doors, Gears, Quarterly_Tax, and Weight. these parameter.

In [22]:
#lets import the libraries
import numpy as np # for mathematical and logical operations and consisting of multidimensional array objects 
import matplotlib.pyplot as plt
import pandas as pd  # for Data Manipulation
import seaborn as sns   #for Advanced visualizations
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_val_score
from collections import Counter
from IPython.core.display import display, HTML
sns.set_style('darkgrid')

In [23]:
#lets load the data set
online = pd.read_csv("ToyotaCorolla.csv", sep=",")
#as the datais not loading properly from the pd function
#we have imported the data from variable explorer

# Data Preprocessing

In [24]:
# Checking the null values, count, columns name and  data describetion  
online.describe()
online.columns
online.count()
online.isna().sum()
online.info

<bound method DataFrame.info of         Id                                              Model  Price  \
0        1      TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13500   
1        2      TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13750   
2        3     �TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13950   
3        4      TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  14950   
4        5        TOYOTA Corolla 2.0 D4D HATCHB SOL 2/3-Doors  13750   
...    ...                                                ...    ...   
1431  1438         TOYOTA Corolla 1.3 16V HATCHB G6 2/3-Doors   7500   
1432  1439  TOYOTA Corolla 1.3 16V HATCHB LINEA TERRA 2/3-...  10845   
1433  1440  TOYOTA Corolla 1.3 16V HATCHB LINEA TERRA 2/3-...   8500   
1434  1441  TOYOTA Corolla 1.3 16V HATCHB LINEA TERRA 2/3-...   7250   
1435  1442        TOYOTA Corolla 1.6 LB LINEA TERRA 4/5-Doors   6950   

      Age_08_04  Mfg_Month  Mfg_Year     KM Fuel_Type   HP  Met_Color  ...  \
0            23         1

In [26]:
#as the first column is given serial number, so this is not required
online = online.drop('Id', axis=1)
online = online.drop('Model',axis=1)
online

Price  Age_08_04  Mfg_Month  Mfg_Year     KM Fuel_Type   HP  Met_Color  \
0     13500         23         10      2002  46986    Diesel   90          1   
1     13750         23         10      2002  72937    Diesel   90          1   
2     13950         24          9      2002  41711    Diesel   90          1   
3     14950         26          7      2002  48000    Diesel   90          0   
4     13750         30          3      2002  38500    Diesel   90          0   
...     ...        ...        ...       ...    ...       ...  ...        ...   
1431   7500         69         12      1998  20544    Petrol   86          1   
1432  10845         72          9      1998  19000    Petrol   86          0   
1433   8500         71         10      1998  17016    Petrol   86          0   
1434   7250         70         11      1998  16916    Petrol   86          1   
1435   6950         76          5      1998      1    Petrol  110          0   

       Color  Automatic  ...  Central_Lock  Powered_Windows  Power_Steering  \
0       Blue          0  ...             1                1               1   
1     Silver          0  ...             1                0               1   
2       Blue          0  ...             0                0               1   
3      Black          0  ...             0                0               1   
4      Black          0  ...             1                1               1   
...      ...        ...  ...           ...              ...             ...   
1431    Blue          0  ...             1                1               1   
1432    Grey          0  ...             0                0               1   
1433    Blue          0  ...             0                0               1   
1434    Grey          0  ...             0                0               0   
1435   Green          0  ...             0                0               1   

      Radio  Mistlamps  Sport_Model  Backseat_Divider  Metallic_Rim  \
0         0          0            0                 1             0   
1         0          0            0                 1             0   
2         0          0            0                 1             0   
3         0          0            0                 1             0   
4         0          1            0                 1             0   
...     ...        ...          ...               ...           ...   
1431      0          1            1                 1             0   
1432      0          0            1                 1             0   
1433      0          0            0                 1             0   
1434      0          0            0                 1             0   
1435      0          0            0                 0             0   

      Radio_cassette  Tow_Bar  
0                  0        0  
1                  0        0  
2                  0        0  
3                  0        0  
4                  0        0  
...              ...      ...  
1431               0        0  
1432               0        0  
1433               0        0  
1434               0        0  
1435               0        0  

[1436 rows x 36 columns]

In [27]:
#lets make the dependant and independant variables
#as from the data set profit is depends on the remaining all
X = online.iloc[:,1:36] #independ variables excluding price 
= online.iloc[:,0]  #depended variable price

In [28]:
#as the State column is catergorical we will convert this using onehot encoding using get_dummies()
#the impact of first NewYork is removed
Fuel_Type= pd.get_dummies(X['Fuel_Type'])
Color = pd.get_dummies(X['Color'])

In [29]:
#lets remove the categorical column
X = X.drop(['Color','Fuel_Type'], axis=1)
X 

Age_08_04  Mfg_Month  Mfg_Year     KM   HP  Met_Color  Automatic    cc  \
0            23         10      2002  46986   90          1          0  2000   
1            23         10      2002  72937   90          1          0  2000   
2            24          9      2002  41711   90          1          0  2000   
3            26          7      2002  48000   90          0          0  2000   
4            30          3      2002  38500   90          0          0  2000   
...         ...        ...       ...    ...  ...        ...        ...   ...   
1431         69         12      1998  20544   86          1          0  1300   
1432         72          9      1998  19000   86          0          0  1300   
1433         71         10      1998  17016   86          0          0  1300   
1434         70         11      1998  16916   86          1          0  1300   
1435         76          5      1998      1  110          0          0  1600   

      Doors  Cylinders  ...  Central_Lock  Powered_Windows  Power_Steering  \
0         3          4  ...             1                1               1   
1         3          4  ...             1                0               1   
2         3          4  ...             0                0               1   
3         3          4  ...             0                0               1   
4         3          4  ...             1                1               1   
...     ...        ...  ...           ...              ...             ...   
1431      3          4  ...             1                1               1   
1432      3          4  ...             0                0               1   
1433      3          4  ...             0                0               1   
1434      3          4  ...             0                0               0   
1435      5          4  ...             0                0               1   

      Radio  Mistlamps  Sport_Model  Backseat_Divider  Metallic_Rim  \
0         0          0            0                 1             0   
1         0          0            0                 1             0   
2         0          0            0                 1             0   
3         0          0            0                 1             0   
4         0          1            0                 1             0   
...     ...        ...          ...               ...           ...   
1431      0          1            1                 1             0   
1432      0          0            1                 1             0   
1433      0          0            0                 1             0   
1434      0          0            0                 1             0   
1435      0          0            0                 0             0   

      Radio_cassette  Tow_Bar  
0                  0        0  
1                  0        0  
2                  0        0  
3                  0        0  
4                  0        0  
...              ...      ...  
1431               0        0  
1432               0        0  
1433               0        0  
1434               0        0  
1435               0        0  

[1436 rows x 33 columns]

In [30]:
#lets use concatination
X = pd.concat([X,Color,Fuel_Type], axis=1)
X 

Age_08_04  Mfg_Month  Mfg_Year     KM   HP  Met_Color  Automatic    cc  \
0            23         10      2002  46986   90          1          0  2000   
1            23         10      2002  72937   90          1          0  2000   
2            24          9      2002  41711   90          1          0  2000   
3            26          7      2002  48000   90          0          0  2000   
4            30          3      2002  38500   90          0          0  2000   
...         ...        ...       ...    ...  ...        ...        ...   ...   
1431         69         12      1998  20544   86          1          0  1300   
1432         72          9      1998  19000   86          0          0  1300   
1433         71         10      1998  17016   86          0          0  1300   
1434         70         11      1998  16916   86          1          0  1300   
1435         76          5      1998      1  110          0          0  1600   

      Doors  Cylinders  ...  Green  Grey  Red  Silver  Violet  White  Yellow  \
0         3          4  ...      0     0    0       0       0      0       0   
1         3          4  ...      0     0    0       1       0      0       0   
2         3          4  ...      0     0    0       0       0      0       0   
3         3          4  ...      0     0    0       0       0      0       0   
4         3          4  ...      0     0    0       0       0      0       0   
...     ...        ...  ...    ...   ...  ...     ...     ...    ...     ...   
1431      3          4  ...      0     0    0       0       0      0       0   
1432      3          4  ...      0     1    0       0       0      0       0   
1433      3          4  ...      0     0    0       0       0      0       0   
1434      3          4  ...      0     1    0       0       0      0       0   
1435      5          4  ...      1     0    0       0       0      0       0   

      CNG  Diesel  Petrol  
0       0       1       0  
1       0       1       0  
2       0       1       0  
3       0       1       0  
4       0       1       0  
...   ...     ...     ...  
1431    0       0       1  
1432    0       0       1  
1433    0       0       1  
1434    0       0       1  
1435    0       0       1  

[1436 rows x 46 columns]

In [31]:
#let's split the data into train and test
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [32]:
#now lets build a model using fitting Multilinear Regression to the training set
from sklearn.linear_model import LinearRegression
start_f = LinearRegression() 
start_f.fit(X_train, y_train)

LinearRegression()

In [33]:
#prediction for the test results
y_pred = start_f.predict(X_test)
y_pred

array([  8902.89094415,  10930.17351334,  13378.01869454,   7677.83247574,
        10432.66576372,   9915.24818761,   6462.49542937,  15332.47255526,
        11479.46751546,  10619.19257447,   6377.88558016,  10074.68400938,
        11118.36064289,   9713.68177557,   7302.89934513,  14768.94772503,
        11436.06948335,   8115.18617751,  11504.16805212,  10862.43148416,
         6885.94236157,   6653.12150519,  11508.87802906,  11798.71454718,
        21631.77943184,   9198.01428745,  13466.87408117,   6033.73766786,
         9283.95849611,   7675.30542709,  10994.43115259,   8381.84334864,
        16348.52189393,   7465.00611289,   9671.861632  ,  11876.21165276,
         9763.45479031,  12415.93153024,  10009.73675029,   8887.69810875,
         8356.28868037,   9952.18199699,  10277.85282611,  15242.64858869,
         9943.24673131,   7259.90101591,  10095.56221832,   6849.91378228,
         8253.41964331,   8672.59746076,  20915.69026066,   9161.48285269,
         7445.49539548,  

In [34]:
#for comarision to check whether the y_test is good or y_pred is good
#we will use r^2 score from the sklearn package
from sklearn.metrics import r2_score
score = r2_score(y_test, y_pred) 
score

0.37407949610941227

In [17]:
# Predicting Cross Validation Score the Test set results
cv_linear = cross_val_score(estimator = start_f, X = X_train, y = y_train, cv = 10)
cv_linear

array([0.85295169, 0.90241947, 0.90174145, 0.91481452, 0.92485328,
       0.9108801 , 0.91970749, 0.82267693, 0.92849608, 0.94058836])

In [18]:
# Predicting R2 Score the Train set results
y_pred_linear_train = start_f.predict(X_train)
r2_score_linear_train = r2_score(y_train, y_pred_linear_train)
r2_score_linear_train

0.9207693782607749

In [19]:
# Predicting R2 Score the Test set results
y_pred_linear_test = start_f.predict(X_test)
r2_score_linear_test = r2_score(y_test, y_pred_linear_test)
r2_score_linear_test

0.37407949610941227

In [20]:
# Predicting RMSE the Test set results
rmse_linear = (np.sqrt(mean_squared_error(y_test, y_pred_linear_test)))
print("CV: ", cv_linear.mean())
print('R2_score (train): ', r2_score_linear_train)
print('R2_score (test): ', r2_score_linear_test)
print("RMSE: ", rmse_linear)

CV:  0.9019129374534772
R2_score (train):  0.9207693782607749
R2_score (test):  0.37407949610941227
RMSE:  2790.6780509855835


Result : By checking the result of model the train score is 92% while train socre is 37% mean our model is under fitting condition is occured 